---
title: "NZSRW kinship analysis"
author: "Robin Aldridge-Sutton"
output:   
  github_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

### Introduction

This is an analysis of the pairwise kinships among a sample of NZ southern right whales using microsatellite genotypes, MtDNA haplotypes, and sex.  It is organised into 6 numbered scripts, and source_scripts.R is used to conveniently call those that are required when changes are made.  It takes multiple csv files as inputs, and writes a single new csv file with integrated data for ~95% of the animals sampled. The data are largely from [Carroll, Emma. (2012) Return of the right whale: Assessment of abundance, population structure and geneflow in the New Zealand southern right whale](http://hdl.handle.net/2292/11541).  The main results are summarised in the other sections of this readme.

For each pair it finds the number of microsatellite loci at which they share one or two alleles, as a simple heuristic for kinship, then it finds pseudolikelihoods given to various simple kinships by all of the data, and the half-sibling versus unrelated pair pseudo-log likelihood ratios.  

It saves pairs which could be POPs according to their observed microsatellite genotypes, and has a shiny app to display them along with one of the years and locations at which they were sampled, but I took it down because I have not been able to distinguish likely POPs so far.

The likelihood calculations could probably be sped up and simplified by finding all possible geno-, haplo-, and sex-pair probabilities just once, and then just indexing them for each observed pair, and by setting probabilities for missing data to 1 when they are no longer needed for adding probabilities over cases.  But these are probably not worth doing unless I end up running simulations with this kind of data.

```{r}
# Load results of analysis
load(file = "rdata/whale_analysis_6.Rdata")
```

### Missing data and dropout for microsatellite genotypes

Locus TR3G1 had the most missing data, and is the only one with significant excess homozygosity.  

```{r}
# Find and show proportions of missing data by locus from largest to smallest
barplot(
  colMeans(missing_ales_2),
  las = 2,
  main = "Missing data by locus",
  ylab = "Proportion"
)
```

```{r}
# Find the excess observed homozygosity at each locus
props_homo_exp <- sapply(ale_dists, function(dist) sum(dist^2))
props_homo_obs <- colMeans(ales_mat_1 == ales_mat_2, na.rm = T)

# Show observed versus expected homozygosity by locus from largest to smallest
# difference
barplot(
  props_homo_obs - props_homo_exp,
  las = 2,
  main = "Excess observed homozygosity",
  ylab = "Proportion of data"
)
```

From Emma's thesis (pg.84):

"Of the 13 loci retained in the dataset, 12 did not deviate significantly from
Hardy-Weinberg equilibrium and showed no signs of stutter, allelic dropout or
null alleles (Table 2.5). The exception was TR3G1; this locus had evidence of
allelic dropout but was retained as it was highly informative and allelic
dropout was accounted for by re-amplifying mismatching loci for suspected
replicate samples."

If we had borderline cases of kinship we could check how this affected them.

### Shared microsatellite alleles, kinship likelihoods, and possible parent-offspring pairs

Parent-offspring pairs (POPs) share an allele by descent at each locus.  This is not enough to distinguish them with these microsatellite genotypes, but we can find the pairs that could possibly be POPs according to them (bearing in mind genotyping error), and check that the more sophisticated analyses below give the same results.

```{r}
# Plot the proportions of loci where pairs are observed to share at least one
# allele
hist(
  ns_loci_ales_shared[ns_loci_both_obs == 13],
  breaks = 100,
  main = "Numbers of loci with at least one shared allele 
  in pairs with complete microsatellite genotypes",
  xlab = "Number of loci",
  ylab = "Number of pairs"
)
```

It is possible to have genotypes that share an allele at a locus but which are still more likely to be unrelated than to be a POP.  For example, if a pair of  microsatellite genotypes are each heterozygous at a locus, and only shares one allele, the conditional probability of the second given that they are unrelated is twice the product of the allele probabilities, and given that they are a POP it is half the unshared allele probability.  So if the probability of the shared allele is greater than one quarter, the unrelated pair prob is higher.  For SNPs, allele probabilities are often greater than one quarter, but you can't have two heterozygous genotypes that only share one allele.  Many of the microsatellite alleles in this dataset have probabilities greater than one quarter, and a few pairs share an allele at every locus but are more likely to be UPs.

Here are the proportions of microsatellite loci where alleles are shared, the likelihoods given to the main simple kinships, and the maximum likelihood kinship, for the first few genopairs, and the first few possible POPs.

```{r}
# Create table of kinship analysis data, including proportions of msat loci with
# 0, 1, and 2 alleles shared, kinship likelihoods, and most likely kinship
pairs_data <- data.frame(
  props_loci_ns_ales_shared,
  round(kin_log_likes_msats, 2),
  MLK = max_like_kinships_msats
)

print(head(pairs_data), row.names = F)
print(pairs_data[head(possible_pop_inds), ], row.names = F)
```

### Kinship log likelihoods based on sexes and haplotypes

Here are the sexes and haplotypes, the likelihoods given by them to the main simple kinships, and the maximum likelihood kinship, for a random sample of genopairs.  When any of the data is missing the likelihoods are ignored although it is possible to calculate them later if necessary. 

```{r}
# Show kinship likelihoods given haplopairs and sexes for random selection of
# animals
pairs_data <- data.frame(
  whale_data[ani_inds_1, 2:3],
  whale_data[ani_inds_2, 2:3],
  round(kin_log_likes_hps_sexes, 2),
  MLK = max_like_kinships_hps_sexes
)
print(pairs_data[sample(n_pairs, 10), ], row.names = F)
```

### Combined kinship log likelihood ratios

The pseudo-log likelihood ratios do not clearly distinguish any kinships for this data, although it may be possible to apply a thresholding technique later.

```{r}
# Plot half-siblings versus unrelated pair plods for pairs with complete data

# Set two plots per page
par(mfrow = c(2, 1))

# Plot full histogram
hist(
  hsp_up_plods_comp, 
  breaks = 200,
  main = "Half-sibling vs unrelated pair PLLRs 
  for pairs with complete data",
  xlab = "PLOD"
)

# Show rare values
hist(
  hsp_up_plods_comp, 
  breaks = 200,
  main = "Large PLODs suggest close-kin",
  xlab = "PLOD",
  ylim = c(0, 20)
)
```